external/js/cothority/src/network/connection.ts

```diff
@@ -242,3 +250,234 @@ export class LeaderConnection extends WebSocketConnection {
+     * @param p active connections
+     */
+    set parallel(p: number) {
+        if (Nodes.parallel > 0) {
+            p = Nodes.parallel;
+        }
+        if (p > this.addresses.length) {
+            this._parallel = this.addresses.length;
+        } else if (p > 0) {
+            this._parallel = p;
+        }
+        while (this._parallel > this.active.length) {
+            this.active.push(this.reserve.shift());
+        }
+        while (this._parallel < this.active.length) {
+            this.reserve.push(this.active.pop());
+        }
+    }
+    // This can be set from the outside to enforce a number of parallel requests. If it
+    // is > 0, it will be enforced.
+    static parallel: number = 0;
+    // the threshold for the ratio of response_node / response_fastest where the node is
+    // put in the 'reserve'.
+    static slowThreshold: number = 10;
+    private readonly addresses: string[];
+    // to the slowest node.
+    private readonly active: Node[] = [];
+    // is the one that was not used for the longest time.
+    private readonly reserve: Node[];
+    // Pool of available connections that are not in the active connections. The first node
+
+    // Current number of nodes in the active queue.
+    private _parallel: number;
+    // Holds the fastest connections for 'parallel' nodes. They are sorted from the fastest
+
+    constructor(r: Roster, parallel: number = 3) {
+        this.addresses = r.list.map((conode) => conode.getWebSocketAddress());
+        shuffle(this.addresses);
+        // Initialize the pool of connections
+        this.reserve = this.addresses.map((addr) => new Node(addr));
+        this.parallel = parallel;
+    }
+
+    /**
+     * Creates a new NodeList for one message.
+     * @param service
+     */
+    newList(service: string): NodeList {
+        if (Nodes.parallel > 0) {
+            this.parallel = Nodes.parallel;
+        }
+        return new NodeList(this, service, this.active.slice(), this.reserve.slice());
+    }
+
+    /**
+     * Marks the node of the given address as having an error. The error must be
+     * a websocket-error 1001 or higher. An error in the request itself (refused
+     * transaction) should be treated as a passing node.
+     * @param address
+     */
+    gotError(address: string) {
+        this.replaceActive(this.index(address));
+    }
+
+    /**
+     * Marks the node with the given address as having successfully treated the
+     * message. It will re-order the nodes to reflect order of arrival. If the
+     * node is more than 10x slower than the fastest node, it will be replaced
+     * with a node from the reserve queue.
+     * @param address node with successful return
+     * @param rang order of arrival
+     * @param ratio delay in answer between first reply and this reply
+     */
+    done(address: string, rang: number, ratio: number) {
+        const index = this.index(address);
+        if (index >= 0) {
```

Comment:
Instead of been a NOP when not found, we can throw an error to show possibly wrong code path.

Reply:
No, because nodes from a previous connection might receive a message very late and have already been replaced by other nodes.

So it's legit that a node calls the `done`, but that it will not change its position, as it's too late.